### PR TITLE
Loki: Log hashed queries emitted by `metrics.go`

### PR DESCRIPTION
**Pull request by DylanGuedes**
**What this PR does / why we need it**:
- Adds a new `query_hash` label pair to queries logged by `metrics.go`. As the name suggests, `query_hash` is just the hashed version of the query being logged.
  - Especially useful when grouping results/behavior of different queries. Ex, you can easily find specific query patterns with `sum by(query_hash)` .

Ex:
```
level=info ts=2023-01-23T14:45:52.743780661Z caller=metrics.go:149 component=frontend org_id=fake latency=fast query="{stream=\"stdout\",pod=\"loki-canary-abcde\"}" query_hash=1360642565 query_type=limited range_type=range length=20s start_delta=58m27.642790018s end_delta=58m7.64279027s step=1s duration=4.761564089s status=200 limit=1000 returned_lines=0 throughput=78kB total_bytes=369kB total_entries=41 store_chunks_download_time=3.037076ms queue_time=1m5.039459998s subqueries=18 cache_chunk_req=3 cache_chunk_hit=3 cache_chunk_bytes_stored=0 cache_chunk_bytes_fetched=2881268 cache_chunk_download_time=2.49811ms cache_index_req=33 cache_index_hit=33 cache_index_download_time=61.548315ms cache_result_req=1 cache_result_hit=0 cache_result_download_time=217.84µs
```

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
N/A

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/logql/metrics.go [hunk 1]
```diff
@@ -108,11 +109,12 @@ func RecordRangeAndInstantQueryMetrics(
 
 	queryTags, _ := ctx.Value(httpreq.QueryTagsHTTPHeader).(string) // it's ok to be empty.
 
-	logValues := make([]interface{}, 0, 20)
+	logValues := make([]interface{}, 0, 21)
```
Comment (dannykopping) [inline pkg/logql/metrics.go:112]: By my count, there are `30` items

**Review by dannykopping — APPROVED** [review]
Great idea @DylanGuedes!

File: pkg/logql/metrics.go (hunk 1 above)
Comment (DylanGuedes) [reply pkg/logql/metrics.go]: Haha your math is better than mine. Fixed!

**Comment by slim-bean** [comment]
We should also add the hash to the "executing query" log lines to make correlating started vs finished queries much easier

File: pkg/logql/engine.go [hunk 2]
```diff
@@ -203,10 +203,11 @@ func (q *query) Exec(ctx context.Context) (logqlmodel.Result, error) {
 	log, ctx := spanlogger.New(ctx, "query.Exec")
 	defer log.Finish()
 
+	queryHash := hashedQuery(q.params.Query())
```
Comment (slim-bean) [inline pkg/logql/engine.go:206]: I think we should verify that the same hash is produced in metrics.go vs this executing query line for queries like:

```
sum(rate(...)) by (label)
```

This query will show up in metrics.go as:

```
sum by (label) (rate(...))
```

And i'm not exactly sure where this re-write happens?

Comment (DylanGuedes) [reply pkg/logql/engine.go:206]: Good catch. I pushed a test that checks whether the hashed query emitted by engine vs metrics.go are equal or not. Apparently, `sum(...) by(label)` resulted in the same query hash in both places so not sure I'm missing something :thinking: 

if you know of other interesting queries for me to test I can add them as new test cases in the scenario I implemented on `engine_test.go`

Comment (slim-bean) [reply pkg/logql/engine.go:206]: ok, no worries! thanks for checking, my recollection here is likely missing something.

**Review by slim-bean — APPROVED** [review]
LGTM!

**Pull request merged**